### ISSUE TITLE
The optimizer is not removing side-effect free constraints

### ISSUE THREAD

**Issue by axic** [Zokrates/ZoKrates#856]
```
struct Memory {
        u8[6] values
}

def main(Memory memory):
        // 56 constraints with this line, or 54 if this is commented out
        bool x = memory.values[1] != 0
        return
```

Of course more pressing in "loops":
```
struct Memory {
        u8[6] values
}

def main(Memory memory):
      for u32 i in 0..5 do
                // 64 constraints with this line, or 54 if this is commented out
                bool x = memory.values[i] != 0
      endfor
      return
```

**Comment by axic** [comment]
Perhaps this shouldn't be an optimiser step, but rather an analyzer warning: "Side-effect free statement."

**Comment by Schaeff** [comment]
This was fixed in #1206
